justfile

```diff
@@ -0,0 +1,109 @@
+# enforce our chosen pip compile flags
+export COMPILE := BIN + "/pip-compile --allow-unsafe"
+
+
+# list available commands
+default:
+    @{{ just_executable() }} --list
+
+
+# clean up temporary files
+clean:
+    rm -rf .venv
+
+
+# ensure valid virtualenv
+virtualenv:
+    #!/usr/bin/env bash
+    # allow users to specify python version in .env
+    PYTHON_VERSION=${PYTHON_VERSION:-python3.10}
+
+    # create venv and upgrade pip
+    test -d $VIRTUAL_ENV || { $PYTHON_VERSION -m venv $VIRTUAL_ENV && $PIP install --upgrade pip; }
+
+    # ensure we have pip-tools so we can run pip-compile
+    test -e $BIN/pip-compile || $PIP install pip-tools
+
+
+_compile src dst *args: virtualenv
+    #!/usr/bin/env bash
+    # exit if src file is older than dst file (-nt = 'newer than', but we negate with || to avoid error exit code)
+    test "${FORCE:-}" = "true" -o {{ src }} -nt {{ dst }} || exit 0
+    {{ COMPILE }} --output-file={{ dst }} {{ src }} {{ args }}
+
+
+# This project currently has no production dependencies, as it vendors them
+# update requirements.prod.txt if requirements.prod.in has changed
+#requirements-prod *args:
+#    {{ just_executable() }} _compile requirements.prod.in requirements.prod.txt {{ args }}
+
+
+# update requirements.dev.txt if requirements.dev.in has changed
+requirements-dev *args: virtualenv
+    {{ just_executable() }} _compile requirements.dev.in requirements.dev.txt {{ args }}
+
+
+# ensure prod requirements installed and up to date
+#prodenv: requirements-prod
+#    #!/usr/bin/env bash
+#    set -eu
+#    # exit if .txt file has not changed since we installed them (-nt == "newer than', but we negate with || to avoid error exit code)
+#    test requirements.prod.txt -nt $VIRTUAL_ENV/.prod || exit 0
+#
+#    $PIP install -r requirements.prod.txt
+#    touch $VIRTUAL_ENV/.prod
+
+
+# && dependencies are run after the recipe has run. Needs just>=0.9.9. This is
+# a killer feature over Makefiles.
+#
+# ensure dev requirements installed and up to date
+devenv: requirements-dev && install-precommit
+    #!/usr/bin/env bash
+    set -eu
+    # exit if .txt file has not changed since we installed them (-nt == "newer than', but we negate with || to avoid error exit code)
+    test requirements.dev.txt -nt $VIRTUAL_ENV/.dev || exit 0
+
+    $PIP install -r requirements.dev.txt
+    touch $VIRTUAL_ENV/.dev
+
+
+# ensure precommit is installed
+install-precommit:
+    #!/usr/bin/env bash
+    BASE_DIR=$(git rev-parse --show-toplevel)
+    test -f $BASE_DIR/.git/hooks/pre-commit || $BIN/pre-commit install
+
+
+# upgrade dev or prod dependencies (specify package to upgrade single package, all by default)
+upgrade env package="": virtualenv
+    #!/usr/bin/env bash
+    opts="--upgrade"
+    test -z "{{ package }}" || opts="--upgrade-package {{ package }}"
+    FORCE=true {{ just_executable() }} requirements-{{ env }} $opts
+
+
+test *args: devenv
+    $BIN/coverage run --module pytest {{ args }}
```

Comment:
We'll also want, on the line below this, `$BIN/coverage report || $BIN/coverage html` otherwise coverage isn't doing very much!

Reply:
So, that line is wrong. As it is, it only runs html if report fails.

I think what you want is `$BIN/coverage report || true; $BIN/coverage html`  - that will run both regardless of reports exit code

Reply:
Hmm, or is it wrong? 
I think you want to exit with the exit code of `coverage report`, but run html anyway?